Unifying the code base behind add_p()

ddsjoberg opened the pull request:
**What changes are proposed in this pull request?**

- The code base powering `add_p()` and related methods has been refactored for better performance, organization, and customizability. (#622)
    - For clarity, a help file listing each test available within gtsummary and the psuedo code for calculating the p-value has been added (`?add_p_test`)
    - Each `add_p()` method now has the `test.args=` argument. Use this argument to pass additional arguments to the statistical method, e.g. `add_p(test = c(age, marker) ~ "t.test", test.args = c(age, marker) ~ list(var.equal = TRUE))`
    - New function was added `all_tests()` to make it easier to select variables based on the test used to calculate the p-value, e.g. `add_p(test = c(age, marker) ~ "t.test", test.args = all_tests("t.test") ~ list(var.equal = TRUE))`
    - Additional tests have been added: paired t-test, signed rank test
    - More robust unit testing implemented for all `add_p()` methods.
- Breaking change. Experimental function `add_p.tbl_survfit(test.args=)` in addition to accepting the formula list notation, also accepted a single string naming a test that was interpretted as `everything() ~ "test_name"`.  The single string is no longer accepted, and users must use the formula notation.
- New function: `all_summary_cols()` selects columns from `tbl_summary`/`tbl_svysummary` object with summary statistics (i.e. "`stat_0`", "`stat_1`", "`stat_2`", etc.)

**If there is an GitHub issue associated with this pull request, please provide link.**
closes #622 
closes #618 

**TO DO**:
Coding:
- [x] Update infrastructure for all `add_p()` functions to have a common format.
- [x] Add an `all_tests()` selector function
- [x] What happens when a test without a footnote annotation is used? Is the footnote NA added?
- [x] Update footnote message to only print for non-missing p-values.
- [x] Add more tests to `add_p.tbl_summary()` like paired tests. For paired tests, add a message when obs are dropped because of no pair. signed-rank test, Friedman's test, McNemar's test, paired t-test
- [x] Can the internal data frame `df_add_p_tests` be moved to an excel file?
- [x] What to do about all these long footnotes?  Delete all references to continuity corrections?

Documentation:
- [x] Document `test.args=` in all functions
- [x] Document all tests.
    - [x] Document all the tests with their names and psuedo code for how the p-value is calculated.
    - [x] Document how to write a custom test....include both methods with and without the footnote message
- [x] Note the breaking change to `test.args=` in `tbl_survfit()`: no longer accepts a string name.
- [x]  Write a good NOTES section listing all changes
- [x] Update `tbl_summary.Rmd` to include example using `test.args=` and `all_tests()`
- [x] Add paired t-test example to gallery

Tests:
- [x] Write tests comparing output to p-values calculated outside of gtsummary
    - [x]  `tbl_summary()`
    - [x] `tbl_svysummary()`
    - [x] `tbl_survfit()`
- [x] Add tests for `all_tests()` selector




--------------------------------------------------------------------------------

Checklist for PR reviewer

- [x] PR branch has pulled the most recent updates from master branch. Ensure the pull request branch and your local version match and both have the latest updates from the master branch.
- [x] If an update was made to `tbl_summary()`, was the same change implemented for `tbl_svysummary()`?
- [x] If a new function was added, function included in `_pkgdown.yml`
- [x] If a bug was fixed, a unit test was added for the bug check
- [x] Run `pkgdown::build_site()`. Check the R console for errors, and review the rendered website.
- [x] Code coverage is suitable for any new functions/features. Review coverage with `covr::report()`. Before you run, set `Sys.setenv(NOT_CRAN=true)` and begin in a fresh R session without any packages loaded. 
- [x] R CMD Check runs without errors, warnings, and notes
- [x] NEWS.md has been updated with the changes from this pull request under the heading "`# gtsummary (development version)`". If there is an issue associated with the pull request, reference it in parantheses at the end update (see NEWS.md for examples).
- [x] `usethis::use_spell_check()` runs with no spelling errors in documentation
- [x] When the branch is ready to be merged into master, increment the version number using `usethis::use_version(which = "dev")`, run `codemetar::write_codemeta()`, approve, and merge the PR. Please use "Suash and merge".

michaelcurry1123 commented:
Create a warning update when test.args are passed but not used. Right now you can pass additional arguments to test.args and they are ignored if the test function does not use them. It might confuse end users if they think the additional argument are going to be used but end up not doing anything. Example below:
```
mod2 <- survfit(Surv(ttdeath, death)~ trt , data = trial)

tbl_survfit(mod2, probs = c(.25,.5,.75)) %>% 
  add_p(test = list(trt ~ "logrank")
        ,test.args = trt ~ list(rho = 0.5)
        )

```

The `test` input specifies logrank, which means rho =0, and will ignore the `test.args` additional input where rho = 0.5. The footnote is correct in saying it is a log-rank test but it is not clear that test.args is ignored.

ddsjoberg commented:
@michaelcurry1123 good points!

1. I updated the tests to print a message if args are mistakenly passed to a test that does not use them.
```r
> tbl_survfit(mod2, probs = c(.25,.5,.75)) %>% 
+   add_p(test = list(trt ~ "logrank"),
+         test.args = trt ~ list(rho = 0.5))
Note for variable 'trt': Argument(s) 'test.args' do not apply and were ignored.
See `?tests` for details.
```

2. I updated the footnote for survdiff tests to print the actual rho value to make the test performed clear.
```r
tbl_survfit(mod2, probs = c(.25,.5,.75)) %>% 
  add_p(test = list(trt ~ "survdiff"),
        test.args = trt ~ list(rho = 0.5))
```

![image](https://user-images.githubusercontent.com/26774684/100883483-616d2a80-347e-11eb-945f-5a37c4c1ee68.png)
